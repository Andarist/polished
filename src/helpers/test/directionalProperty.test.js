// @flow
import directionalProperty from '../directionalProperty'

describe('directionalProperty', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
  })

  it('properly generates properties when passed a hyphenated property', () => {
    expect(directionalProperty('border-width', '12px')).toMatchSnapshot()
  })

  it('properly generates properties when passed a camelCased property', () => {
    expect(directionalProperty('borderWidth', '12px')).toMatchSnapshot()
  })

  it('properly passes just the position when not given a property', () => {
    expect(directionalProperty('', '12px')).toMatchSnapshot()
  })

  // One Param
  it('properly applies a value when passed only one', () => {
    expect(directionalProperty('border', '12px')).toMatchSnapshot()
  })

  // Two Params
  it('properly applies values when passed two', () => {
    expect(directionalProperty('border', '12px', '24px')).toMatchSnapshot()
  })
  it('properly skips top and bottom properties when first value is null', () => {
    expect(directionalProperty('border', null, '12px')).toMatchSnapshot()
  })
  it('properly skips left and right properties when second value is null', () => {
    expect(directionalProperty('border', '12px', null)).toMatchSnapshot()
  })

  // Three Params
  it('properly applies values when passed three', () => {
    expect(
      directionalProperty('border', '12px', '24px', '36px'),
    ).toMatchSnapshot()
  })
  it('properly skips top property when first value is null', () => {
    expect(
      directionalProperty('border', null, '24px', '36px'),
    ).toMatchSnapshot()
  })
  it('properly skips right and left properties when second value is null', () => {
    expect(
      directionalProperty('border', '12px', null, '36px'),
    ).toMatchSnapshot()
  })
  it('properly skips bottom property when last value is null', () => {
    expect(
      directionalProperty('border', '12px', '24px', null),
    ).toMatchSnapshot()
  })

  // Four Params
  it('properly applies values when passed four', () => {
    expect(
      directionalProperty('border', '12px', '24px', '36px', '48px'),
    ).toMatchSnapshot()
  })
  it('properly skips top property when first value is null', () => {
    expect(
      directionalProperty('border', null, '24px', '36px', '48px'),
    ).toMatchSnapshot()
  })
  it('properly skips right property when second value is null', () => {
    expect(
      directionalProperty('border', '12px', null, '36px', '48px'),
    ).toMatchSnapshot()
  })
  it('properly skips bottom property when third value is null', () => {
    expect(
      directionalProperty('border', '12px', '24px', null, '48px'),
    ).toMatchSnapshot()
  })
  it('properly skips left property when fourth value is null', () => {
    expect(
      directionalProperty('border', '12px', '24px', '36px', null),
    ).toMatchSnapshot()
  })

  // Overloaded
  it('throws a warning when passed more than 5 parameters and properly ignores extra parameters', () => {
    expect(
      directionalProperty('border', '10px', '10px', '10px', '10px', '10px'),
    ).toMatchSnapshot()
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })

  // Errors
  it('throws an error when passed a non-string value for property', () => {
    directionalProperty(true, '12px')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
  it('throws an error when passed less than 2 parameters', () => {
    directionalProperty('border')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
})
