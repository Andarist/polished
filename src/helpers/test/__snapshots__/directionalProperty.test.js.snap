exports[`directionalProperty properly applies a value when passed only one 1`] = `
Object {
  "borderBottom": "12px",
  "borderLeft": "12px",
  "borderRight": "12px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly applies values when passed four 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "48px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly applies values when passed three 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "24px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly applies values when passed two 1`] = `
Object {
  "borderBottom": "12px",
  "borderLeft": "24px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly generates properties when passed a camelCased property 1`] = `
Object {
  "borderBottomWidth": "12px",
  "borderLeftWidth": "12px",
  "borderRightWidth": "12px",
  "borderTopWidth": "12px",
}
`;

exports[`directionalProperty properly generates properties when passed a hyphenated property 1`] = `
Object {
  "borderBottomWidth": "12px",
  "borderLeftWidth": "12px",
  "borderRightWidth": "12px",
  "borderTopWidth": "12px",
}
`;

exports[`directionalProperty properly passes just the position when not given a property 1`] = `
Object {
  "bottom": "12px",
  "left": "12px",
  "right": "12px",
  "top": "12px",
}
`;

exports[`directionalProperty properly skips bottom property when last value is null 1`] = `
Object {
  "borderLeft": "24px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips bottom property when third value is null 1`] = `
Object {
  "borderLeft": "48px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips left and right properties when second value is null 1`] = `
Object {
  "borderBottom": "12px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips left property when fourth value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips right and left properties when second value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips right property when second value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "48px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips top and bottom properties when first value is null 1`] = `
Object {
  "borderLeft": "12px",
  "borderRight": "12px",
}
`;

exports[`directionalProperty properly skips top property when first value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "24px",
  "borderRight": "24px",
}
`;

exports[`directionalProperty properly skips top property when first value is null 2`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "48px",
  "borderRight": "24px",
}
`;

exports[`directionalProperty throws a warning when passed more than 5 parameters and properly ignores extra parameters 1`] = `
Object {
  "borderBottom": "10px",
  "borderLeft": "10px",
  "borderRight": "10px",
  "borderTop": "10px",
}
`;

exports[`directionalProperty throws a warning when passed more than 5 parameters and properly ignores extra parameters 2`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- helpers/directionalProperty.js --%c

directionalProperty %cexpects a maximum of %c5 parameters%c. However, you passed %c6 parameters%c. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#directionalProperty%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`directionalProperty throws an error when passed a non-string value for property 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- helpers/directionalProperty.js --%c

directionalProperty %cexpects a parameter of type string. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#directionalProperty%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`directionalProperty throws an error when passed less than 2 parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- helpers/directionalProperty.js --%c

directionalProperty %cexpects a minimum of %c2 parameters%c. However, you passed %c1 parameter%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#directionalProperty%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: red",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
